mindsdb_sql/planner/plan_join.py

```diff
@@ -470,24 +475,82 @@ def process_predictor(self, item, query_in):
 
         # params for model
         model_params = None
-
+        partition_size = None
         if query_in.using is not None:
             model_params = {}
             for param, value in query_in.using.items():
                 if '.' in param:
                     alias = param.split('.')[0]
                     if (alias,) in item.aliases:
                         new_param = '.'.join(param.split('.')[1:])
-                        model_params[new_param] = value
+                        model_params[new_param.lower()] = value
                 else:
-                    model_params[param] = value
+                    model_params[param.lower()] = value
 
-        predictor_step = self.planner.plan.add_step(ApplyPredictorStep(
+            partition_size = model_params.pop('partition_size', None)
+
+        predictor_step = ApplyPredictorStep(
             namespace=item.integration,
             dataframe=data_step.result,
             predictor=item.table,
             params=model_params,
             row_dict=row_dict,
             columns_map=columns_map,
-        ))
-        self.step_stack.append(predictor_step)
+        )
+
+        self.step_stack.append(
+            self.add_plan_step(predictor_step, partition_size=partition_size)
+        )
+
+    def add_plan_step(self, step, partition_size=None):
+        """
+        Adds step to plan
+
+        If partition_size is defined: create partition
+        If partition is active
+            If step can be partitioned:
+                Add step to partition not in plan
+            Otherwise:
+                Add partition to plan
+                Add step to plan
+        """
+        if self.partition:
+            if isinstance(step, (JoinStep, ApplyPredictorStep)):
+                # add to partition
+
+                self.add_step_to_partition(step)
+                return step
+
+        elif partition_size is not None:
+            # create partition
+
+            self.partition = MapReduceStep(
+                values=step.dataframe,
+                reduce='union',
+                step=[],
+                partition=partition_size
+            )
+            self.planner.plan.add_step(self.partition)
+
+            self.add_step_to_partition(step)
+            return step
+
+        else:
+            # next step can't be partitioned.
+            self.close_partition()
+
+        return self.planner.plan.add_step(step)
+
+    def add_step_to_partition(self, step):
+        step.step_num = f'{self.partition.step_num}_{len(self.partition.step)}'
+        self.partition.step.append(step)
+
+    def close_partition(self):
+        # return
+        # if partitions is exist - clear it and replace last stack item with it
+
+        if self.partition:
+            if len(self.step_stack) > 0:
+                self.step_stack[-1] = self.partition
```

Comment:
Minor gripe, maybe it is better to reset the entire stack so that other items are garbage collected? I.e.:

```python
self.step_stack = [self.partition]
```

Reply:
the self.step_stack should be kept, only the last element should be replaced . It could contain other join steps (with next tables or models)

Reply:
Ah, makes sense. Marking as resolved.